presto-geospatial/src/main/java/com/facebook/presto/geospatial/BingTile.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.geospatial;
+
+import com.facebook.presto.spi.PrestoException;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.google.common.base.Preconditions.checkArgument;
+
+public final class BingTile
+{
+    public static final int MAX_ZOOM_LEVEL = 23;
+
+    private final int x;
+    private final int y;
+    private final int zoomLevel;
+
+    private BingTile(int x, int y, int zoomLevel)
+    {
+        checkArgument(zoomLevel <= MAX_ZOOM_LEVEL);
+        this.x = x;
+        this.y = y;
+        this.zoomLevel = zoomLevel;
+    }
+
+    public static BingTile fromCoordinates(int x, int y, int zoomLevel)
+    {
+        return new BingTile(x, y, zoomLevel);
+    }
+
+    public static BingTile fromQuadKey(String quadKey)
+    {
+        int zoomLevel = quadKey.length();
+        checkArgument(zoomLevel <= MAX_ZOOM_LEVEL);
+        int tileX = 0;
+        int tileY = 0;
+        for (int i = zoomLevel; i > 0; i--) {
+            int mask = 1 << (i - 1);
+            switch (quadKey.charAt(zoomLevel - i)) {
+                case '0':
+                    break;
+                case '1':
+                    tileX |= mask;
+                    break;
+                case '2':
+                    tileY |= mask;
+                    break;
+                case '3':
+                    tileX |= mask;
+                    tileY |= mask;
+                    break;
+                default:
+                    throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Invalid QuadKey digit sequence: " + quadKey);
+            }
+        }
+
+        return new BingTile(tileX, tileY, zoomLevel);
+    }
+
+    public int getX()
+    {
+        return x;
+    }
+
+    public int getY()
+    {
+        return y;
+    }
+
+    public int getZoomLevel()
+    {
+        return zoomLevel;
+    }
+
+    public String toQuadKey()
+    {
+        char[] quadKey = new char[this.zoomLevel];
+        for (int i = this.zoomLevel; i > 0; i--) {
+            char digit = '0';
+            int mask = 1 << (i - 1);
+            if ((this.x & mask) != 0) {
```

Comment:
how is this.x = null handled?

Reply:
@souravc83 , I don't think I understand this question. this.x is an int. It can't be null. Would you clarify?

Reply:
Sorry, that was a wrong q (not so familiar with presto). presto already handles nulls I think..